### PR TITLE
support timestamps

### PR DESCRIPTION
**Pull request by xiangxyq**
hi, our project need timestamps with sherpa-ncnn, current sherpa-ncnn has no timestamps output,  I have completed this. But also has some confusions:
1. [sherpa-ncnn-alsa.cc](https://github.com/k2-fsa/sherpa-ncnn/blob/master/sherpa-ncnn/csrc/sherpa-ncnn-alsa.cc) and [sherpa-ncnn-microphone.cc](https://github.com/k2-fsa/sherpa-ncnn/blob/master/sherpa-ncnn/csrc/sherpa-ncnn-microphone.cc) need to print timestamps result
2. https://github.com/xiangxyq/sherpa-ncnn/blob/cj-add-timestamps/sherpa-ncnn/csrc/recognizer.h#L38 
timestamps use float instead int32, if use Convert like sherpa https://github.com/k2-fsa/sherpa/blob/4756d4ec10c61164eff356190740527993844df3/sherpa/cpp_api/online-stream.h#L17 ,  need extra struct
4. https://github.com/xiangxyq/sherpa-ncnn/blob/28b77f7f6f1820fb0149530b5ef5c933b7275fb2/sherpa-ncnn/csrc/greedy-search-decoder.cc#L43 , here why  push blanck_id to token, how to handle timestamps?  this question also in https://github.com/k2-fsa/sherpa-ncnn/blob/b80b74695be0eafcc7a3815a5647899b7b9a82b5/sherpa-ncnn/csrc/modified-beam-search-decoder.cc#L120

Thanks!

### CODE REVIEW

File: sherpa-ncnn/csrc/hypothesis.h [hunk 1]
```diff
@@ -33,7 +33,7 @@ struct Hypothesis {
 
   // timestamps[i] contains the frame number after subsampling
   // on which ys[i] is decoded.
-  std::vector<int32_t> timestamps;
+  std::vector<float> timestamps;
```
Comment (csukuangfj) [inline sherpa-ncnn/csrc/hypothesis.h:36]: Is there a reason to use float here?

File: sherpa-ncnn/csrc/recognizer.cc [hunk 2]
```diff
@@ -85,7 +86,16 @@ void Recognizer::AcceptWaveform(float sample_rate, const float *input_buffer,
...
-RecognitionResult Recognizer::GetResult() { return decoder_->GetResult(); }
+RecognitionResult Recognizer::GetResult() {
+  auto r = decoder_->GetResult();
+
+  float frame_shift_s = fbank_opts_->frame_opts.frame_shift_ms / 1000.;
+  for (auto &t : r.timestamps) {
+    t *= frame_shift_s;
+  }
+
+  return std::move(r);
```
Comment (csukuangfj) [inline sherpa-ncnn/csrc/recognizer.cc:97]:
```suggestion
  return r;
```

File: sherpa-ncnn/csrc/recognizer.h [hunk 3]
```diff
@@ -114,6 +127,7 @@ class Recognizer {
 
  private:
   std::unique_ptr<Model> model_;
+  std::unique_ptr<knf::FbankOptions> fbank_opts_;
```
Comment (csukuangfj) [inline sherpa-ncnn/csrc/recognizer.h:130]:
```suggestion
  knf::FbankOptions fbank_opts_;
```

File: sherpa-ncnn/csrc/recognizer.cc [hunk 4]
```diff
@@ -85,7 +86,16 @@ void Recognizer::AcceptWaveform(float sample_rate, const float *input_buffer,
 
 void Recognizer::Decode() { decoder_->Decode(); }
 
-RecognitionResult Recognizer::GetResult() { return decoder_->GetResult(); }
+RecognitionResult Recognizer::GetResult() {
+  auto r = decoder_->GetResult();
+
+  float frame_shift_s = fbank_opts_->frame_opts.frame_shift_ms / 1000.;
+  for (auto &t : r.timestamps) {
+    t *= frame_shift_s;
```
Comment (csukuangfj) [inline sherpa-ncnn/csrc/recognizer.cc:94]:
```suggestion
    t *= frame_shift_s * 4;
```

You can assume that the subsampling factor is 4.

**Review by csukuangfj — CHANGES_REQUESTED** [review]

**Comment by csukuangfj** [comment]
Could you use
https://github.com/k2-fsa/sherpa-ncnn/blob/master/scripts/check_style_cpplint.sh
to fix your code style?

**Comment by csukuangfj** [comment]
> [sherpa-ncnn-alsa.cc](https://github.com/k2-fsa/sherpa-ncnn/blob/master/sherpa-ncnn/csrc/sherpa-ncnn-alsa.cc) and [sherpa-ncnn-microphone.cc](https://github.com/k2-fsa/sherpa-ncnn/blob/master/sherpa-ncnn/csrc/sherpa-ncnn-microphone.cc) need to print timestamps result

Could you support these two files as well?

**Comment by xiangxyq** [comment]
> Could you use https://github.com/k2-fsa/sherpa-ncnn/blob/master/scripts/check_style_cpplint.sh to fix your code style?

ok

**Comment by xiangxyq** [comment]
> Thanks for your contribution!
> 
> > , here why push blanck_id to token, how to handle timestamps? this question also in
> 
> The reason is to make the following code easier to write
> 
> https://github.com/k2-fsa/sherpa-ncnn/blob/b80b74695be0eafcc7a3815a5647899b7b9a82b5/sherpa-ncnn/csrc/greedy-search-decoder.cc#L31-L36
> 
> If we put blanks at the beginning, we don't need to test whether the vector `result_.tokens` is empty or not

i see, will lead tokens size() != timestamps size(),  remove blank token in Recognizer::GetResult()?

https://github.com/xiangxyq/sherpa-ncnn/blob/0f30fb825f1a81018168014f036ec4f8a15cb786/sherpa-ncnn/csrc/recognizer.cc#L89-L98

**Comment by csukuangfj** [comment]
@xiangxyq 
Thanks again for your contribution. I am merging it and am also refactoring the code.

**Comment by xiangxyq** [comment]
@csukuangfj some suggest has not to do, will continue to do.

**Comment by csukuangfj** [comment]
> @csukuangfj some suggest has not to do, need continue.

Please wait before I finish refactoring the code.

Thanks!

**Pull request merged**